lib/Config.php

```diff
@@ -251,44 +251,67 @@ public function getTurnServers(): array {
 	}
 
 	/**
-	 * Generates a username and password for the TURN server
+	 * Prepares a list of TURN servers with username and password
 	 *
 	 * @return array
 	 */
 	public function getTurnSettings(): array {
 		$servers = $this->getTurnServers();
 
 		if (empty($servers)) {
-			return [
+			return [[
 				'schemes' => '',
 				'server' => '',
 				'username' => '',
 				'password' => '',
 				'protocols' => '',
-			];
-		}
-
-		// For now we use a random server from the list
-		try {
-			$server = $servers[random_int(0, count($servers) - 1)];
-		} catch (\Exception $e) {
-			$server = $servers[0];
+			]];
 		}
 
 		// Credentials are valid for 24h
 		// FIXME add the TTL to the response and properly reconnect then
 		$timestamp = $this->timeFactory->getTime() + 86400;
 		$rnd = $this->secureRandom->generate(16);
 		$username = $timestamp . ':' . $rnd;
-		$password = base64_encode(hash_hmac('sha1', $username, $server['secret'], true));
-
-		return [
-			'schemes' => $server['schemes'],
-			'server' => $server['server'],
-			'username' => $username,
-			'password' => $password,
-			'protocols' => $server['protocols'],
-		];
+
+		foreach ($servers as $server) {
+			$password = base64_encode(hash_hmac('sha1', $username, $server['secret'], true));
+
+			if (!empty($server['server'])) {
+				$schemes = explode(',', $server['schemes']);
+				$protocols = explode(',', $server['protocols']);
+				foreach ($schemes as $scheme) {
+					foreach ($protocols as $proto) {
+						$turnservers[] = [
+							'schemes' => $scheme,
+							'server' => $server['server'],
+							'username' => $username,
+							'password' => $password,
+							'protocols' => $proto,
+						];
+					}
+				}
+			}
+
+			// Prepare three random TURN servers (turn-udp, turn-tcp, turns-tcp)
+			foreach ($turnservers as $key => $val) {
+				if ($val['schemes'] === 'turn' && $val['protocols'] === 'udp') {
+					$turn_udp[] = $key;
+				}
+				if ($val['schemes'] === 'turn' && $val['protocols'] === 'tcp') {
+					$turn_tcp[] = $key;
+				}
+				if ($val['schemes'] === 'turns' && $val['protocols'] === 'tcp') {
+					$turns_tcp[] = $key;
+				}
+			}
+
+			(!empty($turn_udp)) ? $turnSettings[] = $turnservers[$turn_udp[array_rand($turn_udp, 1)]] : '';
+			(!empty($turn_tcp)) ? $turnSettings[] = $turnservers[$turn_tcp[array_rand($turn_tcp, 1)]] : '';
+			(!empty($turns_tcp)) ? $turnSettings[] = $turnservers[$turns_tcp[array_rand($turns_tcp, 1)]] : '';
```

Comment:
Should a `turns_udp` added aswell for the rare cases someone specifies only a turnserver with `turns` and `udp` (dtls over udp)?